v3.0.1 documentation edits (#3755)
apollographql/apollo-kotlin: docs/source/testing/mocking-responses.mdx

---
title: Mocking responses
---

> Apollo Kotlin provides two tools to mock responses in tests:

Author: If this is too long we can also split in 2 pages (with cross references). LMK what you think :)

Reviewer: Yep, I'd be in favour of splitting.

- `TestNetworkTransport`: allows to specify the GraphQL responses returned by `ApolloClient`
- `MockServer`: a low-level HTTP server that can be used to mock any responses

In most cases, `TestNetworkTransport` will be more straightforward to use in your tests, while `MockServer` can be

> useful to test specific server behavior, such as error cases, custom headers, timeouts, etc.

Reviewer: Do we want to be opinionated there? I'm personally the kind of person that'd prefer `MockServer` because this touched as little of the production code as possible (basically just changes the `serverUrl`). Maybe we can gather some feedback what's deemed easiest/more flexible, etc... before making a recommendation?

Author: Totally agree. Will use your suggestion below.

> In most cases, `TestNetworkTransport` will be more straightforward to use in your tests, while `MockServer` can be
> useful to test specific server behavior, such as error cases, custom headers, timeouts, etc.

Reviewer: Maybe something like:

```suggestion
`MockServer` creates a full HTTP server. It is useful to test specific server behaviours, such as error cases, HTTP headers, timeouts, etc. It will also require very little changes to your production code as you'll only need to change `serverUrl`.
`TestNetworkTransport` will replace the HTTP call with a preregistered GraphQL response. It is more lightweight and requires more modifications to your production code.
```

Basically changing `"straighforward"` to `"lightweight"`

## `TestNetworkTransport`

This is a high level test API which allows to specify the responses that will be returned by your `ApolloClient`.

Add the dependency to your project's `build.gradle` file:

```kotlin:title=build.gradle[.kts]
dependencies {
  testImplementation("com.apollographql.apollo3:apollo-testing-support:x.y.z")
}
```

Enable the `TestNetworkTransport` by passing it to the `ApolloClient` builder:

```kotlin{2}
val apolloClient = ApolloClient.Builder()
    .networkTransport(TestNetworkTransport())
    .build()
```

> Either keep a reference to the `TestNetworkTransport` instance, or use the `testNetworkTransport` extension function to
> retrieve it:

Reviewer: I'd recommend using the extension function like for `ApolloClient.apolloStore`. From my experience this is what most users do.

```suggestion
Use the `testNetworkTransport` extension function to retrieve the `TestNetworkTransport`:
```

```kotlin
val testNetworkTransport = apolloClient.testNetworkTransport
```

You can then use the `register` function to specify the responses that will be returned in function of the received
operations:

```kotlin{7}

> val query = GetHeroQuery("001")

Reviewer: ```suggestion
val testQuery = GetHeroQuery("001")
```

val testData = GetHeroQuery.Data {
  hero = droidHero {
    name = "R2D2"
  }
}

> apolloClient.testNetworkTransport.register(query, testData)
> 
> val actual = apolloClient.query(query).execute().data!!

Reviewer: ```suggestion
apolloClient.testNetworkTransport.register(testQuery, testData)

val actual = apolloClient.query(testQuery).execute().data!!
```

assertEquals(testData.hero.name, actual.hero.name)
```

The `register` function accepts an `Operation` (here, the query) and the `ApolloResponse` to be returned when it is
executed. As a shortcut you can pass a `Data` directly instead of an `ApolloResponse`, like in the example above.

> Note how you can use the [Test builders](./test-builders/) to easily create the data.

Reviewer: Life is not easy :P. 

Joking asides, "easy" is a very relative concept so different readers might have different expectations. I'd replace with "type-safe", which is objectively true and the same for everyone

```suggestion
Note how you can use the [Test builders](./test-builders/) to create the data in a type-safe way.
```

Author: Totally agree 🙏 Trying to avoid this kind of terms, but they slip 😅

### Advanced usage

In some cases it can be useful to have different responses returned when executing the same operations consecutively.

To that effect, instead of mapping operations to responses with `register`, it is possible to **enqueue** responses:
they will be returned in the order they were added. To do so, pass a `QueueTestNetworkTransportHandler` to

> the `TestNetworkTransport`, and use its `enqueue` function:

Reviewer: I'd keep the symmetry with MockServer advanced usage and make this section primarily about defining a custom `TestNetworkTransportHandler`

Reading that doc, I wonder if we should even push the symmetry further and have both `TestNetworkTransport` and `MockServer` use a queue by default. That will make it easier to learn one coming from the other. Or (thinking out loud) have both `Map` and `Queue` options by detault:

```kotlin
class TestNetworkTransport(handler: TestNetworkTransportHandler = QueueTestNetworkTransportHandler) {

  fun enqueue(response: ApolloResponse) // throws if handler is not QueueTestNetworkTransportHandler
  fun register(operation: Operation, response: ApolloResponse) // throws if handler is not RegistryTestNetworkTransportHandler
}

class MockServer(handler: MockServerHandler) {

  fun enqueue(response: HttpResponse)
  fun register(request: HttpRequest, response: HttpResponse)
}

```

Author: Good point!

On the MockServer side, I didn't make an equivalent `MapMockServerHandler` at all (not 100% sure it would be useful, and raises a few questions about how `MockRequest` equality should be implemented - I'll have one more look at it).

For `TestNetworkTransport`, I felt `MapTestNetworkTransportHandler` was a sensitive default but tbh I didn't think too much about it :). In the end it'd be nice if the default reflects on "how most users use it", but we don't really know, and I agree that having the same default as the MockServer is a plus (yay consistency!). I'll change the default to use Queue handler, and also add the `enqueue` method (and keep the register one), and update the doc.

While `TestNetworkTransport` functions at the GraphQL level, `MockServer` resides at the HTTP level and can be used to
mock arbitrary HTTP responses.

Add the dependency to your project's `build.gradle` file:

```kotlin:title=build.gradle[.kts]
dependencies {
  testImplementation("com.apollographql.apollo3:apollo-mockserver:x.y.z")
}
```

Here is how to use it:

```kotlin
// Create a mock server
val mockServer = MockServer()

// Use its URL on your ApolloClient
val apolloClient = ApolloClient.Builder().serverUrl(mockServer.url()).store(store).build()

// Enqueue responses
mockServer.enqueue("""{"data": {"random": 42}}""")
mockServer.enqueue(
    MockResponse(
        body = "Internal server error",
        statusCode = 500,
        headers = mapOf("X-Test" to "true"),
        delayMillis = 1000L,
    )
)

// Execute queries
val response1 = apolloClient
    .query(GetRandomQuery())
    .execute()

val response2 = apolloClient
    .query(GetRandomQuery())
    .execute()

// Don't forget to stop the server when you're done
mockServer.stop()
```

Note how the `enqueue` function normally takes a `MockResponse` as a parameter, while a shorthand version taking
a `String` also exists.

> Also note how a delay can be passed to the `MockResponse` constructor to simulate network latency.

Reviewer: Put this as comments in the sample code? By the time I reached this, I'm too lazy to scroll back 😅
